File: time_test/satellite.library/cpp/strings.cpp
// cpp/strings.cpp -- the C++ equivalent of programs/strings.satl: a string joined to itself four
// times and compared with a second join, 300,000 turns. Not line for line: it first prints the
// blank line satl prints after its dashes. The strings are std::u16string because satl's are
// 16-bit characters. same and counter never pass 300,000, so long long holds them.
//
// BUILD: clang++ -std=c++20 -O2 strings.cpp -o strings      (g++ works the same)

#include <cstdio>
#include <string>

int main()
{
    std::printf("\n");

    std::u16string piece = u"the satellite programming language races itself, compiled twice over.";
    std::u16string joined = u"";
    long long same = 0;
    long long counter = 0;
    while (counter < 300000) {
        joined = piece + piece + piece + piece;
        if (joined == piece + piece + piece + piece) {
            same = same + 1;
        }
        counter = counter + 1;
    }
    std::printf("%lld\n", same);
    return 0;
}

File: time_test/satellite.library/python/strings.py
# time_test/satellite.library/python/strings.py -- the Python equivalent of
# programs/strings.satl: a string joined to itself and compared, 300,000 turns.
# A line-for-line translation. Python keeps these ASCII strings at one byte a character where
# satl uses 16-bit characters.
# The blank line printed first is the one satl prints after its line of dashes, so the answers
# compare byte for byte.


def main():
    piece = "the satellite programming language races itself, compiled twice over."
    joined = ""
    same = 0
    counter = 0
    while counter < 300000:
        joined = piece + piece + piece + piece
        if joined == piece + piece + piece + piece:
            same = same + 1
        counter = counter + 1
    print(same)
    return


print()
main()
